recipe/meta.yaml

```
@@ -12,51 +15,77 @@ source:
-    - boost-cpp
-    - cython
-    - numpy 1.16.*
-    - python
-    - setuptools
-    - setuptools_scm
-    - six
+outputs:
+  - name: pyarrow
+    script: build-pyarrow.sh  # [not win]
+    script: bld.bat           # [win]
+    version: {{ version }}
+    build:
+      number: {{ number }}
+      string: "py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ number }}_{{ build_ext }}"
+      skip: true              # [cuda_compiler_version not in (undefined, "None", "9.2")]
+      track_features:
+        {{ "- arrow-cuda" if cuda_enabled else "" }}
+      activate_in_script: True
 
-  run:
-    - {{ pin_compatible('numpy', lower_bound='1.16') }}
-    - boost-cpp
-    - futures  # [py==27]
-    - glog
-    - pandas
-    # empty parquet-cpp metapackage, force old versions to be uninstalled
-    - parquet-cpp 1.5.1.*
-    - python
-    - six
-    - snappy
+    requirements:
+      build:
+        - wget                    # [cuda_compiler_version is not undefined and cuda_compiler_version == '9.2']
+        - cmake
+        - ninja
+        - make                # [unix]
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        # pyarrow does not require nvcc but it needs to link against libraries in arrow-cpp=*=*cuda
+        - {{ compiler("cuda") }}  # [cuda_compiler_version is not undefined and cuda_compiler_version == '9.2']
+      host:
+        # directly pin boost-cpp as we also seem to directly include boost symbols
+        # in the Python modules.
+        - arrow-cpp {{ version }}
+        - arrow-cpp-proc * {{ build_ext }}
+        - boost-cpp
+        - cython
+        - numpy >=1.16
+        - python
+        - setuptools
+        - setuptools_scm
+        - six
+      run:
+        - {{ pin_compatible('numpy', lower_bound='1.16') }}
+        - boost-cpp
+        - arrow-cpp {{ version }}
+        - futures  # [py==27]
+        - glog
+        - pandas
+        # empty parquet-cpp metapackage, force old versions to be uninstalled
+        - parquet-cpp 1.5.1.*
+        - python
+        - six
+        - snappy
+      run_constrained:
+        - arrow-cpp-proc * {{ build_ext }}
 
-test:
-  imports:
-    - pyarrow
-    - pyarrow.dataset  # [not py==27]
-    - pyarrow.flight   # [not py==27]
-    - pyarrow.gandiva  # [(not py==27) and (not (aarch64 or ppc64le))]
-    - pyarrow.orc      # [unix]
-    - pyarrow.parquet
-    - pyarrow.plasma   # [unix]
-    - pyarrow.fs
-    - pyarrow._s3fs
-    - pyarrow._hdfs
+    test:
+      imports:
+        - pyarrow
+        - pyarrow.dataset  # [not py==27]
+        - pyarrow.flight   # [not py==27]
+        - pyarrow.gandiva  # [(not py==27) and (not (aarch64 or ppc64le))]
+        - pyarrow.orc      # [unix]
+        - pyarrow.parquet
+        - pyarrow.fs
+        - pyarrow._s3fs
+        - pyarrow._hdfs
+        - pyarrow.plasma   # [unix and cuda_compiler_version in (undefined, "None")]
+        # CUDA dependent pyarrow.plasma and pyarrow.cuda are tested in build-arrow.sh
```

IMO this should be tested here as well, what works in the build environment does not necessarily work in a user install (which is what the test environment reproduces).
```suggestion
        - pyarrow.plasma   # [unix]
        - pyarrow.cuda     # [unix and cuda_compiler_version != "None"]
```

---

I agree, in general. However, in this particular case (as well as any CUDA enable conda package), the test environment is a subset of user environment as the user is responsible for providing working CUDA environment.

In addition, CUDA 9.2 that conda-forge uses for build, does not have `cuda-compat` libraries that are required for running these import tests. To fix this issue, one needs a time machine..

---

I think it could be solved with a newer `linux-anvil-cuda` docker image for 9.2 from conda-forge, IIUC?

---

I'm happy to do this myself, but I'm still trying to figure out what would be necessary.

---

`linux-anvil-cuda` docker image for 9.2 is based on `nvidia/cuda:9.2-devel-centos6` and I cannot see that this image has the needed changes. So, I doubt that an update will resolve the issue.

---

@h-vetinari , your help would be highly appreciated as this is a blocker to many packages.

---

> `linux-anvil-cuda` docker image for 9.2 is based on `nvidia/cuda:9.2-devel-centos6` and I cannot see that this image has the needed changes. So, I doubt that an update will resolve the issue.

I mean, we can do whatever we want in the `linux-anvil-cuda` image with a PR, so whatever is missing from `nvidia/cuda:9.2-devel-centos6` could be added?

> @h-vetinari, your help would be highly appreciated as this is a blocker to many packages.

Same here, but I'm an exceedingly poor choice of champion for this (no GPU infra / package building experience). I'm piecing some things together as I go, but more often than not, I've been dependent on the insights of @isuruf and @jakirkham.